File: examples/www/public/assets/scripts/sound.lua
local Sound = {}
Sound.__index = Sound

function Sound.load(path)
  local handle = load_sound(path)
  if handle == nil or handle == 0 then
    return nil
  end

  return setmetatable({
    handle = handle,
  }, Sound)
end

function Sound:play()
  if self.handle ~= nil and self.handle ~= 0 then
    play_sound(self.handle)
  end
end

function Sound:destroy()
  if self.handle ~= nil and self.handle ~= 0 then
    destroy_sound(self.handle)
    self.handle = 0
  end
end

return Sound

File: examples/remote/server/src/sound.ts
import type { ResourceManager } from "./resource_manager";
import {
  rl_sound_create,
  rl_sound_destroy,
  rl_sound_play,
  set_resource_manager as set_sound_resource_manager,
} from "./rl/rl_sound";

export class Sound {
  handle: number | null = null;
  private dirty = false;

  constructor(
    public readonly path: string,
    private readonly resourceManager: ResourceManager,
  ) {}

  static async load(resourceManager: ResourceManager, path: string): Promise<Sound> {
    const sound = new Sound(path, resourceManager);
    set_sound_resource_manager(resourceManager);
    sound.handle = await rl_sound_create(path);
    return sound;
  }

  play(): void {
    if (this.handle == null) {
      return;
    }

    this.dirty = true;
  }

  sync(): boolean {
    if (this.handle == null || !this.dirty) {
      return false;
    }

    const ok = rl_sound_play(this.handle);
    if (ok) {
      this.dirty = false;
    }

    return ok;
  }

  async destroy(): Promise<void> {
    if (this.handle == null) {
      return;
    }

    set_sound_resource_manager(this.resourceManager);
    await rl_sound_destroy(this.handle);
    this.handle = 0;
    this.dirty = false;
  }
}
